### PR TITLE
Rewrite printf functionality

### PR DESCRIPTION
**Pull request by quinnj**
Based on https://github.com/JuliaLang/julia/pull/32799.

The main commits unique here rewrite the internal `base/printf.jl` functionality for about half the code, with 2-3x perf improvements. It utilizes the new ryu float printing mechanics in https://github.com/JuliaLang/julia/pull/32799. These commits also remove all the grisu code.

Status: this is pretty far along, I haven't done the `%a`/`%A` type specifier yet, or made it work for `BigFloat` either, and there's still a good chunk of tests we need, but it's pretty useable and worth starting to get reviews on.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
You may already have fixed this, but note https://github.com/JuliaLang/julia/issues/16977.

**Comment by StefanKarpinski** [comment]
This code is so impressively bonkers 😂. Thank you for doing this, @quinnj!

**Comment by musm** [comment]
wow great work. Does this  make it any easier to implement a non-macro `printf` function ?

**Comment by quinnj** [comment]
@StefanKarpinski, I agree :). It's interesting to go back and look at all this code that was written so early on, like a peek into an "old world" of sorts. I think it's a big testament to how much the compiler and code generation "mental model" has improved: I think reaching for meta-programming used to be so much more necessary in order to hand-craft code, whereas today you can rely so much on the compiler that you can just write the code much more "naturally" and everything just works.

@musm, yes, the approach in this PR basically follows what we do for `Dates` formatting or `Regex`: i.e. you can create a "static" `Printf.Format(str)` Format object, with a corresponding `Printf.format"%d"` macro string to create a Format at parse-time. We then have a `Printf.format(io, f::Format, args...)` method that applies the formatting of `f::Format` to `args...`. So the `@printf` macro is really provided here for backwards compatibility, since the definition is essentially just:
```julia
macro printf(io, f, args...)
    fmt = Printf.Format(f)
    return esc(:(Printf.format($io, $fmt, $(args...)))
end
```

File: contrib/generate_precompile.jl
Comment (KristofferC) [inline contrib/generate_precompile.jl]: Why disable all precompile in the printf stdlib?

Comment (quinnj) [reply contrib/generate_precompile.jl]: Ah, I was going to mention it to you, it seems we run into an unfortunate bug w/ how literally we're generating the precompile statements:
```julia
Generating precompile statements...┌ Error: Failed to precompile precompile(Tuple{typeof(Base.Printf.format), Array{UInt8, 1}, Int64, Base.Printf.Format{Base.CodeUnits{UInt8, String}, Tuple{Base.Printf.Spec{Base.Val{Char(0x64000000)}}, Base.Printf.Spec{Base.Val{Char(0x73000000)}}, Base.Printf.Spec{Base.Val{Char(0x73000000)}}}}, Float64, String, String})
└ @ Main.anonymous /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:167
ERROR: LoadError: LoadError: Base.CodePointError{UInt32}(0x64000000)
Stacktrace:
 [1] code_point_err(::UInt32) at ./char.jl:86
 [2] Char(::UInt32) at ./char.jl:160
 [3] top-level scope at string:1
 [4] include_string(::Module, ::String, ::String) at ./loading.jl:1064
 [5] include_string at ./loading.jl:1068 [inlined]
 [6] macro expansion at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:165 [inlined]
 [7] macro expansion at ./util.jl:212 [inlined]
 [8] (::getfield(Main.anonymous, Symbol("##2#6")){Float64,Base.DevNull})(::String, ::IOStream) at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:159
 [9] mktemp(::getfield(Main.anonymous, Symbol("##2#6")){Float64,Base.DevNull}, ::String) at ./file.jl:577
 [10] mktemp at ./file.jl:575 [inlined]
 [11] generate_precompile_statements() at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:70
 [12] top-level scope at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:180
in expression starting at string:1
in expression starting at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:8
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [/Users/jacobquinn/juliasrc/usr/lib/julia/sys-o.a] Error 1
make: *** [julia-sysimg-release] Error 2
```
I also added the other check because I ran into:
```julia
Generating precompile statements...┌ Error: Failed to precompile precompile(Tuple{typeof(Base.io_has_tvar_name), Base.IOContext{REPL.Terminals.TTYTerminal}, Symbol, Type{Vararg{Any, N} where N}})
└ @ Main.anonymous /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:171
ERROR: LoadError: LoadError: TypeError: in Type, in parameter, expected Type, got Vararg{Any,N} where N
Stacktrace:
 [1] top-level scope at string:1
 [2] include_string(::Module, ::String, ::String) at ./loading.jl:1064
 [3] include_string at ./loading.jl:1068 [inlined]
 [4] macro expansion at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:169 [inlined]
 [5] macro expansion at ./util.jl:212 [inlined]
 [6] (::getfield(Main.anonymous, Symbol("##2#6")){Float64,Base.DevNull})(::String, ::IOStream) at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:163
 [7] mktemp(::getfield(Main.anonymous, Symbol("##2#6")){Float64,Base.DevNull}, ::String) at ./file.jl:577
 [8] mktemp at ./file.jl:575 [inlined]
 [9] generate_precompile_statements() at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:70
 [10] top-level scope at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:184
in expression starting at string:1
in expression starting at /Users/jacobquinn/juliasrc/contrib/generate_precompile.jl:8
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [/Users/jacobquinn/juliasrc/usr/lib/julia/sys-o.a] Error 1
make: *** [julia-sysimg-release] Error 2
```

Comment (quinnj) [reply contrib/generate_precompile.jl]: The first precompile statement errors because `Char(0x64000000)` throws an error, but it should just be `Char('d')`

Comment (KristofferC) [reply contrib/generate_precompile.jl]: https://github.com/JuliaLang/julia/pull/32610 should fix the Char problem.

Blacklisting the whole stdlib seems a bit heavy handed. Can't we just list the methods that fail?

Comment (quinnj) [reply contrib/generate_precompile.jl]: I mean, the whole stdlib is `@printf` and `@sprintf`, which lower to `Printf.format`, so......no, there aren't any other methods that fail, because there's only one. Can we expedite #32610?

**Comment by quinnj** [comment]
Just a status update here: tests are all passing, so I feel like #32799 and this PR could be merged at any time. Happy to add a NEWS entry if people would like, though there should be no user-facing changes. The one awkwardness in this PR is the skipping of precompiling for the Printf module, since there seems to be some lower/precompile-statement-generation bug, but it sounds like there might be a fix for that.

File: contrib/generate_precompile.jl
Comment (KristofferC) [reply contrib/generate_precompile.jl]: Oh, I see. If #32610 fixes the whole thing it would be good to just merge that. Otherwise, I guess we can just blacklist the module but this feels like a method that would be quite nice to precompile so we should try not to forget about it.

**Comment by quinnj** [comment]
Given Stefan's [comment](https://github.com/JuliaLang/julia/pull/32799#issuecomment-521003743), I'll plan on merging this (which includes #32799) in 24 hours, unless anyone else wants me to hold off for review or other concerns.

**Comment by vtjnash** [comment]
Looking at this somewhat quickly, I'm a bit more concerned about the implementation of this and leaves me wondering what we should do before the freeze. I'd be somewhat interested to know if the compile-time dispatch is essential, because that seems to be adding a lot of complexity here and performance cliffs here. It also looked liked you had to hard-code all of the type behaviors so that it's not really extensible anymore.

Finally, is anyone else currently extending Printf.jl that'll be broken by this? While I get that it was hard to extend the old code, it was possible, so someone might be depending on that, but I just don't know. We could move the existing Printf (and Grisu) code inside the printf module and move it to Compat to preserve that API as a down-grade version option as a preparatory step.

**Comment by jmkuhn** [comment]
DecFP.jl extends [Base.Printf.fix_dec](https://github.com/JuliaMath/DecFP.jl/blob/58325feffe77037c87c650d743e6aed1e8c326fd/src/DecFP.jl#L286-L325) and [Base.Printf.ini_dec](https://github.com/JuliaMath/DecFP.jl/blob/58325feffe77037c87c650d743e6aed1e8c326fd/src/DecFP.jl#L327-L353) to allow `printf` formatting.  I haven't looked at the details of this PR closely enough to determine how `printf` formatting could be implemented for DecFP after this PR.

**Comment by quinnj** [comment]
Yeah, it looks like ArbNumerics.jl, QuadMath.jl, DoubleFloats.jl, and DecFP.jl are the only packages that extended Printf. I'll look into their usages to see what we can do for compat here.

**Comment by quinnj** [comment]
Quick update on Printf package extenders:
* ArbNumerics.jl just converts to BigFloat, so that's easy
* QuadMath.jl calls it's own internal `quadmath_snprintf` function, which mirrors how we handle BigFloats, so also not very hard to cover
* DoubleFloats.jl just convert to their respective Float cousin (i.e. `Double64` => `Float64`), so also easy
* DecFP.jl is more tricky, since they're directly implementing the `Printf.ini_` routines. I'm not sure what the best thing to do here though. Maybe we can include a little helper that takes the `ini_` outputs and does the formatting for you. 

@vtjnash, I looked into removing the conversion specifier dispatching, but the result wasn't very satisfying. You end up with `formatints`, `formatstrings`, `formatfloats`, etc. functions that basically mirror the dispatched versions. The performance hit on not dispatching was about 20-30% slower, but for some reason even slower for floats (I didn't quite get to the bottom of why). Also, the partial loop unrolling that I assume you're referring to w/ the "performance cliffs" remark is orthogonal to the conversion specifier type parameter. The loop unrolling helps because your signature is essentially `printf(fmt, args...)`, and with heterogenously typed `args`, just looping over them causes dynamic dispatch. In the case I've included, we partially unroll up to 8 args, which I would assume probably covers 99% of cases, and it makes a huge difference performance-wise to have each arg type-stably formatted.

So all-in-all, I'm inclined to keep the dispatch mechanics since it's a nice performance boost, while the "world" of possible types is still pretty small.

**Comment by quinnj** [comment]
Ok, I've cleaned things up a bit more and added a new `Printf.tofloat` function. This would allow things like:
```julia
# for DoubleFloats.jl
Printf.tofloat(x::Double64) = Float64(x)
Printf.tofloat(x::Double32) = Float32(x)
Printf.tofloat(x::Double16) = Float16(x)

# for ArbNumerics.jl
Printf.tofloat(x:: ArbFloat) = BigFloat(x)
```
For QuadMath.jl, I modified the `string(f::Spec; modifier="")` function so you can pass in a custom modifier (like we do for BigFloats), so for QuadMath, it could simply define:
```julia
function Printf.fmt(buf, pos, arg::Float128, spec::Spec{T}) where {T <: Floats}
    ptr = pointer(buf, pos)
    newpos = ccall((:quadmath_snprintf, : libquadmath), Cint, (Ptr{UInt8}, Csize_t, Ptr{UInt8}, Cfloat128), ptr, length(buf), string(spec; modifier="Q"), d)
    return pos + newpos
end
```
So those both seem like good solutions. I'll keep noodling on the DecFP.jl case.

**Comment by quinnj** [comment]
The 32-bit windows build failed w/ something about reinterpretarray.jl or higherorderfns.jl crashes? Have we seen anything like that before? The freebsd build failed on something related to libgit2. It doesn't seem like either of those issues are related to changes here, but speak up if you think otherwise.

**Comment by quinnj** [comment]
Ok, I slept on the DecFP.jl situation and I think I have a plan that could work well, it looks something like:
* Include the `fix_dec` and `ini_dec` functions in Printf for package extension; this allows DecFP not to break w/ these updates
* Provide a `Printf.digits(x, hex)` method in Printf for more general extension; the API is you overload `Printf.digits(x::MyNumericType, hex::Bool)` and return a buffer w/ digit chars for your numeric type, along w/ where the decimal point should be printed (pretty much a combined version of `fix_dec`, `ini_dec`, `fix_hex`, and `ini_hex`); the fallback definition would call `ini_dec`, `fix_dec`, etc.
* Provide a fallback `fmt` function that would first call `digits(x, hex)`, and then apply the various spacing/padding/flag operations on the returned digits.

We already have most of this logic lying around, so I don't think it will be too much lift to add it in and get it working.

**Comment by quinnj** [comment]
Ok, I pushed some compat code in the form of `fmtfallback` that calls the `ini_dec` and `fix_dec` code that DecFP.jl calls. I've confirmed that DecFP tests all pass w/ this code; so we now have a compat plan in place for every known extender of Printf.

**Comment by vtjnash** [comment]
> Also, the partial loop unrolling that I assume you're referring to w/ the "performance cliffs" remark is orthogonal to the conversion specifier type parameter

That, and also the dispatch penalty of putting all of this into type-parameters computed from runtime values. Just doesn't seem like a great trade-off.

**Comment by quinnj** [comment]
@StefanKarpinski, do you still want to review this before merging? Looks like it needs a rebase, but I'm happy to merge if you don't think you'll be able to get to it any time soon.

**Comment by KristofferC** [comment]
> That, and also the dispatch penalty of putting all of this into type-parameters computed from runtime values. Just doesn't seem like a great trade-off.

I don't think this should block from merge this unless there is some actual measurements that show this is a real problem. To me, this looks like a big improvement over the status quo, and we can always iterate on the details later.

**Review by vtjnash — CHANGES_REQUESTED** [review]
address review comments and benchmark requests

**Comment by quinnj** [comment]
> That, and also the dispatch penalty of putting all of this into type-parameters computed from runtime values. Just doesn't seem like a great trade-off.

But for the most part, they aren't true "runtime" values, since we call `Printf.Format(fmt_str)` at macro expansion time. And we've ***required*** the format string to be constant up until now anyway. This approach also just mirrors how we do `dateformat"yyyy-mm-dd"` formatting, which I think has worked out fairly well. I think it's reasonable that in most cases, you know your format string ahead of time, so creating the slightly more typed format type statically to have better runtime printing performance does seem like a good tradeoff, and one we haven't heard any complaints about from the date formatting approach.

Also, did you see [my comment](https://github.com/JuliaLang/julia/pull/32859#issuecomment-521451972)? You can take a look at the type-parameter-less approach [here](https://github.com/quinnj/Prints.jl/blob/jq/notypeparm/src/Prints.jl); and as I mentioned in that comment, the performance hit was 20-30%, so non-trivial. Just wanted to make sure you saw that as I did spend quite a bit of time/effort going with no type parameters, but after trying out both approaches, I do feel like including the type parameters is worth it.

I'll rebase this PR after https://github.com/JuliaLang/julia/pull/33313 is merged (which I think is ready to go).

**Comment by quinnj** [comment]
Ok, a bit hairy, but finally got this rebased. @StefanKarpinski, did you still want to look over this? Happy to find some time we could video chat or real-time slack if that would be easier; or we can merge and run.

**Comment by StefanKarpinski** [comment]
Oh man, I should probably look but finding the time is going to be tough. Give me until end of week and if I haven't done it by then, we can just yolo this and review later.

**Comment by vtjnash** [comment]
I still had some design concerns that the code is pushing things to type parameters that don’t need to be there (well below our typical threshold for inlining) and then not using dispatch. Also, looks like this code has a hardcoded implementation for Big numbers, which is unfortunate. I won’t really have time to say more specific fixes for either thing though, even just trying to run the various versions you’ve done well to prepare and compare.

I think the reason we want to replace the current implementation is to make it less complex and more predictable, so I’m not sure this does either thing. But I know that just sounds like FUD. There’s the other repo you have above which does seem better on both of those axes, so it’s also maybe a question of which one we want to keep in stdlib—or if we just excise all three of them at this point and let users pick.

**Comment by StefanKarpinski** [comment]
Yeah, maybe let's go a little slow on this one. I'll find some time to look through.

**Comment by quinnj** [comment]
@StefanKarpinski, do you think you'll find time to review this? Or should we merge? I'm happy to jump on a quick video chat if that would help the review process.

**Comment by StefanKarpinski** [comment]
Just got back from vacation, let's do a hangout next week.

**Comment by davidanthoff** [comment]
I'd love to see this make it into 1.4!

**Comment by StefanKarpinski** [comment]
https://discourse.julialang.org/t/julia-1-4-feature-freeze-dec-15th/30453

**Comment by davidanthoff** [comment]
Well, then 1.5 :)

**Comment by quinnj** [comment]
Hey, I rebased this! @StefanKarpinski, still interested in reviewing? @simonbyrne, care to take a look?

**Comment by StefanKarpinski** [comment]
I will be looking at it, but that should not dissuade @simonbyrne also since this is a meaty enough change to warrant more than one set of eyes on it.

File: stdlib/Printf/src/Printf.jl
Comment (giordano) [inline stdlib/Printf/src/Printf.jl:361]: Merge conflict?

**Comment by vtjnash** [comment]
Can we get https://github.com/quinnj/Prints.jl/blob/jq/notypeparm/src/Prints.jl into PR-form? The info above said that can get within 30% of this performance at a small fraction of the compile and memory cost.

**Comment by StefanKarpinski** [comment]
I was not aware of Prints.jl! That's actually much closer to how I imagined rewriting this.

**Comment by quinnj** [comment]
I mean.....it's just more work to switch over and honestly the memory/compile times are not _that_ significant, especially considering the performance cost. As I've mentioned before, the approach in this PR is the exact same framework/approach we already take for date formatting/parsing, so it's just a little befuddling why we wouldn't be ok with that same approach here. I know in my own, sole use-case for this functionality (csv writing), a 30% perf hit would be a deal-breaker; if I need to write more than several thousand floats out to a file, the little extra compile time quickly pays for itself, and is a huge win once we get to millions/billions of floats written.

**Comment by StefanKarpinski** [comment]
Ok, will review as-is then.

**Comment by KristofferC** [comment]
Aren't we already now doing effectively the same as in this PR in that we "specialize" on the formatting string by expanding the macro?

**Comment by quinnj** [comment]
> Aren't we already now doing effectively the same as in this PR in that we "specialize" on the formatting string by expanding the macro?

Yes, and much more inefficiently, since it's all expression generation instead of using a type parameter).

**Comment by vtjnash** [comment]
True, but currently we support more than 8 arguments without a performance penalty, and support arbitrary custom types instead of just a hardcoded list. Also, why would doing something as poorly as before be a design goal, if you've already shown how to do it more flexibly also?

**Comment by quinnj** [comment]
> Also, why would doing something as poorly as before be a design goal, if you've already shown how to do it more flexibly also?

Performance. In case you didn't see my last post.

File: test/worlds.jl
Comment (quinnj) [inline test/worlds.jl:309]: @timholy, can you comment on the use of `Printf.gen` here? Is there another method we can call to get similar results? I'd like to get rid of `Printf.gen` in the rewrite, so any suggestion to change this would be great.

**Comment by quinnj** [comment]
Opened https://github.com/JuliaLang/julia/issues/37387 for the single linuxaarch64 failing case.

**Comment by quinnj** [comment]
Ok, this has been open long enough I decided to become another independent reviewer 😄 ; but seriously, it was actually good to go back over the proposed code with relatively "fresh eyes". I did some cleanup, added some comments, and fixed a few minor things. I think this is in pretty good shape, and will plan on merging in a day or two unless someone else would like to request a review in the mean time.

File: stdlib/Printf/src/Printf.jl
Comment (vtjnash) [inline stdlib/Printf/src/Printf.jl:379]: This needs to not be hardcoded here (especially since Keno has just proposed removing BigFloat)

File: stdlib/Printf/test/runtests.jl
Comment (vtjnash) [inline stdlib/Printf/test/runtests.jl:327]: TODO?

File: stdlib/Printf/src/Printf.jl
Comment (quinnj) [reply stdlib/Printf/src/Printf.jl:379]: What exactly shouldn't be hard-coded? The library name? If we move BigFloat to stdlib, wouldn't Printf just take a dependency on it?

Comment (simonbyrne) [reply stdlib/Printf/src/Printf.jl:379]: why not make it a function interface which we can intercept? e.g. so that we could extend it for Quadmath.jl or DecFP.jl

File: stdlib/Printf/src/Printf.jl
Comment (vtjnash) [inline stdlib/Printf/src/Printf.jl:271]: `AbstractFloat` and `Integer` aren't the only type classes that can exist. Other user types may need the chance to override this too.

File: stdlib/Printf/src/Printf.jl
Comment (vtjnash) [reply stdlib/Printf/src/Printf.jl:379]: There shouldn't be a list of known types in this module—that's why Julia has dispatch

Comment (quinnj) [reply stdlib/Printf/src/Printf.jl:379]: Please see previous comments [here](https://github.com/JuliaLang/julia/pull/32859#issuecomment-521512416) and [here](https://github.com/JuliaLang/julia/pull/32859#issuecomment-521839355) for the compat/intercept story. In short, there are already ways to overload/provide for custom types to hook in to Printf at various levels. And we're at least providing solutions to all known users/overloaders of current Printf code, in most cases allowing for a much simpler integration.

File: stdlib/Printf/src/Printf.jl
Comment (quinnj) [reply stdlib/Printf/src/Printf.jl:271]: You'll have to be more concrete with what you'd like changed here, since it's not very clear from your comment. If were to guess, you're imagining we'd want something similar to `Printf.tofloat` for integers, so having a `Printf.tointeger` that custom types could overload if they'd like a custom type formatted as an integer. I thought of that, but the interface for an "integer" is also much simpler than floats; i.e. if your custom type supports multiplication, division, and `ndigits`, then it'll "just work". This is one of those cases where if someone has a valid use-case, it'd be pretty easy to add, but no use pre-engineering if not necessary.

**Comment by vtjnash** [comment]
I'll also reiterate my request to remove some of the complexity and compiler heft from the initial PR (ala https://github.com/quinnj/Prints.jl/blob/jq/notypeparm/src/Prints.jl). I'm not advising we should never do it, but I'm suggesting the default thing in Base should try to be compact and clear and fast to compile (because I expect it to make a difference on time-to-first-plot), and then we should add complexity and other knobs as follow up work.

File: stdlib/Printf/src/Printf.jl
Comment (vtjnash) [reply stdlib/Printf/src/Printf.jl:379]: Yes, so this should be dog-fooding whatever capability it has. I think possibly there's a mismatch here on implementation vs policy, where this function is confused as to which it is doing?

I've added the needs-docs tag, since that comment (on how to integrate this with a custom type) would be good to have in the PR.

File: stdlib/Printf/src/Printf.jl
Comment (vtjnash) [reply stdlib/Printf/src/Printf.jl:271]: Yeah, that might help? Currently this only handles half of the cases, just in Base:
```
julia> subtypes(Real)
4-element Vector{Any}:
 AbstractFloat
 AbstractIrrational
 Integer
 Rational
```
Which doesn't always work now:
```
julia> @printf "%d" π
ERROR: MethodError: no method matching Integer(::Irrational{:π})
julia> @printf "%d" 2//4
ERROR: InexactError: Integer(1//2)
```
But sometimes it does:
```
julia> @printf "%d" 2//1
2
```

**Comment by quinnj** [comment]
> I'll also reiterate my request to remove some of the complexity and compiler heft from the initial PR

I'll also reiterate my position: the performance penalty of not using type parameters for format specifiers makes this PR worthless for my own desired use-case: allowing users to provide format strings for printing when writing csv files with `CSV.write`. It's just an application that is very highly scrutinized for performance, and when writing millions and millions of floats, it makes a huge overall difference in performance, so the extra compilation cost is more than made up for in faster printing.

If others agree we shouldn't use type parameters, I'll close this PR and just put this code in a package, but I don't think I have the energy/motivation to update this PR with the other approach. If someone else wants to take that on, I'd be happy to review/give directions to what already exists in the Prints.jl repo.

Given there's only been a single voice of concern there, however, I'll still plan on merging this soon, since there have been more comments in favor of merging.

File: stdlib/Printf/src/Printf.jl
Comment (quinnj) [reply stdlib/Printf/src/Printf.jl:379]: It is? Up on line 371, it calls `x = tofloat(arg)`, which is one of the documented ways to hook a custom type into printf integration.

Comment (vtjnash) [reply stdlib/Printf/src/Printf.jl:379]: If true, there wouldn't be a branch here naming a random external type as `x isa BigFloat`. Why is that present?

Comment (quinnj) [reply stdlib/Printf/src/Printf.jl:379]: Maybe look more closely at the code? The logic is as follows:
* Call `x = tofloat(arg)`, which is documented such that overloaders need to return one of `Float16`, `Float32`, `Float64`, or `BigFloat`. 
* If `tofloat` returned a `BigFloat` (as `BigFloat` does itself), and is finite, we call libmpfr to format it
* If the BigFloat is not finite, (i.e. `NaN`, `Inf`, or `-Inf`), we convert to `Float64` to then be formatted

Does that make sense? I might be missing a valid concern here, so please explain in more detail if so.

Comment (simonbyrne) [inline stdlib/Printf/src/Printf.jl:379]: I may be wrong, but I think @vtjnash was thinking something along these lines:

```suggestion
tofloat(x) = Float64(x)
tofloat(x::Base.IEEEFloat) = x
tofloat(x::BigFloat) = x

@inline function fmt(buf, pos, arg, spec::Spec{T}) where {T <: Floats}
    fmt(buf, pos, tofloat(arg), spec::Spec{T}) where {T <: Floats}
end

@inline function fmt(buf, pos, x::BigFloat, spec::Spec{T}) where {T <: Floats}
    if !isfinite(x)
         return fmt(buf, pos, Float64(x), spec)
    end
    leftalign, plus, space, zero, hash, width, prec =
    spec.leftalign, spec.plus, spec.space, spec.zero, spec.hash, spec.width, spec.precision
    ptr = pointer(buf, pos)
    newpos = @ccall "libmpfr".mpfr_snprintf(ptr::Ptr{UInt8}, (length(buf) - pos + 1)::Csize_t, string(spec; modifier="R")::Ptr{UInt8}; arg::Ref{BigFloat})::Cint
    newpos > 0 || error("invalid printf formatting for BigFloat")
    return pos + newpos
end
@inline function fmt(buf, pos, x::Base.IEEEFloat, spec::Spec{T}) where {T <: Floats}
    leftalign, plus, space, zero, hash, width, prec =
    spec.leftalign, spec.plus, spec.space, spec.zero, spec.hash, spec.width, spec.precision

```

Comment (quinnj) [reply stdlib/Printf/src/Printf.jl:379]: Meh, I'm not a huge fan of splitting methods just for dispatch when a trivial `isa` allows the same compilation benefits.

File: stdlib/Printf/test/runtests.jl
Comment (quinnj) [reply stdlib/Printf/test/runtests.jl]: Fixed

File: test/worlds.jl
Comment (timholy) [reply test/worlds.jl:309]: Sorry I missed this ping. It might have been stale, it doesn't look coupled to anything else.

I hadn't followed this, but it looks like awesome work. Many thanks!

**Comment by DilumAluthge** [comment]
I think this PR might have broken ProgressMeter: https://github.com/timholy/ProgressMeter.jl/issues/167

It looks like the offending line in the ProgressMeter source code might be [this line](https://github.com/timholy/ProgressMeter.jl/blob/4752da3e6c1fe10faea12eeb365ca81d82db66a2/src/ProgressMeter.jl#L236).

Is this something that needs to be fixed in Julia or in the ProgressMeter package?

**Comment by martinholters** [comment]
Seems to be a bug in the new printf code:
```julia
julia> @sprintf "%%%s" "a" # ok
"%%a"

julia> @sprintf "%s%%%s" "a" # should require two args
"a%%%s"

julia> @sprintf "%s%%%s" "a" "b" # should produce "a%%b"
ERROR: ArgumentError: mismatch between # of format specifiers and provided args: 1 != 2
```

**Comment by quinnj** [comment]
Ok, fix for problem w/ escaped `'%'` is up: https://github.com/JuliaLang/julia/pull/37470.

**Pull request merged**
